### PR TITLE
Add the server selection screen from EI

### PR DESCRIPTION
**Pull request by pixlwave**
Part of #40, this PR adds the server selection screen (with no network checks on the server address for now).

As this first commit is already large enough, I'll open a separate PR to enable UI tests which I'm currently working on by adding an AuthenticationService along with a mock implementation.

| Light Mode | Dark Mode |
| - | - |
| Simulator Screen Shot - iPhone 13 mini - 2022-06-29 at 16 40 43 | Simulator Screen Shot - iPhone 13 mini - 2022-06-29 at 16 41 04 |

### CODE REVIEW

File: ElementX/Sources/Screens/Authentication/LoginScreen/LoginCoordinator.swift [hunk 1]
```diff
@@ -178,7 +178,34 @@ final class LoginCoordinator: Coordinator, Presentable {
...
+        let coordinator = ServerSelectionCoordinator(parameters: parameters)
+        coordinator.callback = { [weak self, weak coordinator] result in
+            guard let self = self, let coordinator = coordinator else { return }
+            self.serverSelectionCoordinator(coordinator, didCompleteWith: result)
+        }
+        
+        coordinator.start()
+        add(childCoordinator: coordinator)
+        
+        let modalRouter = NavigationRouter(navigationController: UINavigationController())
```
Comment (ismailgulek) [inline ElementX/Sources/Screens/Authentication/LoginScreen/LoginCoordinator.swift:193]: Can we use an `ElementNavigationController` instance here?

File: ElementX/Sources/Screens/Authentication/ServerSelection/ServerSelectionCoordinator.swift [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    }
+    
+    /// Updates the login flow using the supplied homeserver address, or shows an error when this isn't possible.
+    private func useHomeserver(_ homeserverAddress: String) {
+        startLoading()
+        
+        let homeserverAddress = LoginHomeserver.sanitized(homeserverAddress)
+        callback?(.selected(LoginHomeserver(address: homeserverAddress)))
+        
+        stopLoading()
```
Comment (ismailgulek) [inline ElementX/Sources/Screens/Authentication/ServerSelection/ServerSelectionCoordinator.swift:108]: Does it make more sense to stop loading first and then calling the callback?

File: ElementX/Sources/Screens/Authentication/ServerSelection/View/ServerSelectionScreen.swift [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    // MARK: - Properties
+    
+    // MARK: Private
+    
+    @FocusState var isTextFieldFocused: Bool
+    
+    // MARK: Public
+    
+    @ObservedObject var viewModel: ServerSelectionViewModel.Context
```
Comment (ismailgulek) [inline ElementX/Sources/Screens/Authentication/ServerSelection/View/ServerSelectionScreen.swift:29]: Optional but maybe we can name this as `context` as we did for other views. Apparently i forgot to mention this for your other PRs though 🤦‍♂️, so it's up to you.

File: changelog.d/40.change [hunk 4]
```diff
@@ -1 +1 @@
-Add a the splash screen and login screen from Element iOS along with a UserSessionStore.
+Add a the splash, login and server selection screens from Element iOS along with a UserSessionStore.
```
Comment (ismailgulek) [inline changelog.d/40.change:1]:
```suggestion
Add the splash, login and server selection screens from Element iOS along with a UserSessionStore.
```

File: ElementX/Sources/Screens/Authentication/ServerSelection/MockServerSelectionScreenState.swift [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+enum MockServerSelectionScreenState: CaseIterable {
+    case matrix
+    case emptyAddress
+    case invalidAddress
+    case nonModal
+    
+    /// Generate the view struct for the screen state.
+    var viewModel: ServerSelectionViewModel {
+        // swiftlint:disable:next implicit_getter
+        @MainActor get {
```
Comment (ismailgulek) [inline ElementX/Sources/Screens/Authentication/ServerSelection/MockServerSelectionScreenState.swift:28]:
```suggestion
    @MainActor var viewModel: ServerSelectionViewModel {
           ...
```
It seems there is no obstacle to make the property `@MainActor` and get rid of the getter.

**Review by ismailgulek — APPROVED** [review]
Minor comments inline, otherwise LGTM!

File: ElementX/Sources/Screens/Authentication/LoginScreen/LoginCoordinator.swift (hunk 1 above)
Comment (pixlwave) [reply ElementX/Sources/Screens/Authentication/LoginScreen/LoginCoordinator.swift:193]: Ah yeah thats the one! I reached for `RiotNavigationController` but it wasn't there 🙈

File: ElementX/Sources/Screens/Authentication/ServerSelection/View/ServerSelectionScreen.swift (hunk 3 above)
Comment (pixlwave) [reply ElementX/Sources/Screens/Authentication/ServerSelection/View/ServerSelectionScreen.swift:29]: Oh yeah, I intended to do this right from the start but missed it for all 3 screens 👌

File: ElementX/Sources/Screens/Authentication/ServerSelection/MockServerSelectionScreenState.swift (hunk 5 above)
Comment (pixlwave) [reply ElementX/Sources/Screens/Authentication/ServerSelection/MockServerSelectionScreenState.swift:28]: Questioning why I even thought to reach for this style now 🤔

**Pull request merged**